### PR TITLE
New Bot: Doctor Repulsor

### PR DESCRIPTION
**Pull request by asamy**
Hi!

Bot name & based upon what hero
Doctor Repulsor (Doctor Repulsor)

Is the bot the first for a hero, change in behaviour, significant change to a present bot
First.

General overview of the bots intent
The bot can mimic a lot of human-like things, e.g. Ult -> proc E attack -> Ult and also W -> Q -> attack -> ult -> attack, etc.

Any significant implementations to be highlighted
Nope

What Members of the Community Bot Team were consulted in the creation of this bot
References from Naib were used: http://forums.heroesofnewerth.com/sh...ation-Tutorial

Open discussion link
http://forums.heroesofnewerth.com/showthread.php?566438-Doctor-Repulsor&p=16358959#post16358959
## 

So, here's source code for my doctor repulsor, I've had fun watching him VS Dampeer for an hour and it was really an awesome experience developing this Bot.

P.S.: Sorry for the commit disorder, apparently, I had a time leap.
P.S.: This is only one hour work (basically both development and learning since I just started) which means I will be improving this quite a lot in the upcoming days.  But feel free to merge it and I will create new pull requests.

Thanks!

### CODE REVIEW

**Comment by asamy45** [comment]
Sorry for the commit disorder, apparently, I had a time leap.

File: bots/heroes/dr_main.lua [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+behaviorLib.LaneItems  = {"Item_Bottle", "Item_Marchers", "Item_Steamboots", "Item_GraveLocket", "Item_Lightbrand"}
+behaviorLib.MidItems  = {"Item_GrimoireOfPower", "Item_Silence"}
+behaviorLib.LateItems  = {"Item_Protect", "Item_Morph"}
+
+-- skillbuild table, 0=q, 1=w, 2=e, 3=r, 4=attri
+object.tSkills = {
+	0, 2, 1, 0,		-- Q E W Q	(4)
+	1, 3, 0, 0,		-- W R Q Q	(8)		Maxed Q here	(Level 1 ult)
+	1, 3, 1, 2,		-- W R W E  (12)	Maxed W here	(Level 2 ult)
+	2, 2, 2, 3,		-- E E E R  (16)					(Level 3 ult)
```
Comment (Anakonda) [inline bots/heroes/dr_main.lua]: You picked E 5times

File: bots/heroes/dr_main.lua [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+------------------------------
+-- @param: none
+-- @return: none
+function object:SkillBuild()
+	local unitSelf = self.core.unitSelf
+	if skills.abilQ == nil then
+		skills.abilQ = unitSelf:GetAbility(0)
+		skills.abilW = unitSelf:GetAbility(1)
+		skills.abilR = unitSelf:GetAbility(3)
+		skills.abilAttributeBoost = unitSelf:GetAbility(4)
```
Comment (Anakonda) [inline bots/heroes/dr_main.lua]: You don't need Attributes

File: bots/heroes/dr_main.lua [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+	end
+
+	local nlev = unitSelf:GetLevel()
+	local nlevpts = unitSelf:GetAbilityPointsAvailable()
+	for i = nlev, nlev+nlevpts do
+		unitSelf:GetAbility( object.tSkills[i] ):LevelUp()
+	end
+end
+
+local function funcFindItemsOverride(botBrain)
```
Comment (Anakonda) [inline bots/heroes/dr_main.lua]: Find items is obsolete - see core.GetItem

File: bots/heroes/dr_main.lua [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+object.FindItemsOld = core.FindItems
+core.FindItems = funcFindItemsOverride
+
+-- @param: tGameVariables
+-- @return: none
+function object:onthinkOverride(tGameVariables)
+	self:onthinkOld(tGameVariables)
+
+	-- If we're low HP, run for our life!
+	local unitSelf = core.unitSelf
```
Comment (Anakonda) [inline bots/heroes/dr_main.lua]: This needs to be moved to behaviorLib.CustomRetreatExecute
Also you always move to same direction even when on hellbourne

File: bots/heroes/dr_main.lua [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+-- bonus agression points that are applied to the bot upon successfully using a skill/item
+object.nContrpUse = 30
+object.nOpposingUse = 15
+object.nSpeedUse = 35
+object.nHFlowerUse = 40
+object.nMorphUse = 40
+
+--thresholds of aggression the bot must reach to use these abilities
+object.nContrpThreshold = 25
+object.nOpposingThreshold = 10
```
Comment (Anakonda) [inline bots/heroes/dr_main.lua]: Thresholds under 20 do nothing

File: bots/heroes/dr_main.lua [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+					local range = skills.abilW:GetRange()
+					if nTargetDistanceSq < (range * range) then
+						bActionTaken = core.OrderAbilityEntity(botBrain, skills.abilW, unitTarget)
+
+						-- if we can cast Q, do it too
+						local qRange = skills.abilQ:GetRange()
+						if skills.abilQ:CanActivate()
+							and (unitSelf:GetLevel() >= 5 or nTargetDistanceSq < (qRange * qRange))
+						then
+							core.OrderAbility(botBrain, skills.abilQ)
```
Comment (Anakonda) [inline bots/heroes/dr_main.lua]: Use bActionTaken here

File: bots/heroes/dr_main.lua [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+				core.OrderAbilityPosition(botBrain, skills.abilQ, vecMyPosition)
+			else
+				core.OrderMoveToUnitClamp(botBrain, unitSelf, unitTarget)
+			end
+		end
+	end
+
+	if not bActionTaken then
+		return object.harassExecuteOld(botBrain)
+	end
```
Comment (Anakonda) [inline bots/heroes/dr_main.lua]: At the end return bActionTaken

File: bots/heroes/dr_main.lua (hunk 6 above)
Comment (asamy) [inline bots/heroes/dr_main.lua]: I don't think it's nessecary here, we already assigned to it before for the W, and that was the main purpose, Q is an extra.

File: bots/heroes/dr_main.lua (hunk 4 above)
Comment (asamy) [inline bots/heroes/dr_main.lua]: Actually on Hellbourne it's fine, on Legion it's the problem, yep.

File: bots/heroes/dr_main.lua (hunk 6 above)
Comment (Anakonda) [inline bots/heroes/dr_main.lua]: In that case you should note that per call for HarassHeroExecuteOverride only the last order gets executed.

Comment (asamy) [inline bots/heroes/dr_main.lua]: Ah, alright.  That explains why it was kinda delayed, so it seemed that the other thing got executed a tick later.  I'll remove that behavior.

Comment (asamy) [inline bots/heroes/dr_main.lua]: Okay, I think it should be pretty fine now, can you review the latest changes?

**Comment by asamy** [comment]
Is this going to have any comments soon?  I have never in my entire life gotten such delayed response for an open source contribution.

I mean if you'd rather merge broken bots and even fixing them before merge: http://i.imgur.com/3pl1ux7.png which is quite ridiculous

**Comment by Anakonda** [comment]
I'm sorry that you feel wounded even thou it is justified.
In scale of how fast bots are reviewed yours is going fast. Reasons are both Frostburn newer had much intrest in bots and community team losing theirs.

**Comment by Anakonda** [comment]
Ill do it tonight.

File: bots/heroes/dr_main.lua [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+	-- If we're low HP, run for our life!
+	local unitSelf = core.unitSelf
+	if unitSelf and unitSelf:GetHealth() > 0 and unitSelf:GetHealth() < 400
+		and skills.abilR:CanActivate() and unitSelf:GetMana() > 230
+	then
+		-- stolen from RallyTest and adjusted a bit to fit
+		local vecOrigin = unitSelf:GetPosition()
+		local vecDirection = Vector3.Create(1, 0)
+		vecDirection = core.RotateVec2D(vecDirection, unitSelf:GetTeam() == 2 and 90 or -90)
+		core.OrderAbilityPosition(self, skills.abilR, vecOrigin + vecDirection * 500)
```
Comment (Anakonda) [inline bots/heroes/dr_main.lua]: self is nil here, should be botBrain.
Also could use behaviorLib.GetSafeBlinkPosition(vecDesiredPosition, nRange) to find a position that is not 'stuck'
position would be well - core.allyWell:GetPosition() and range what ever you like.
Ill merge after that.

**Comment by Anakonda** [comment]
Oh you'll get free avatar - anything but war effort revards. Please post your wish in the forum submission thread.

File: bots/heroes/dr_main.lua (hunk 8 above)
Comment (asamy) [inline bots/heroes/dr_main.lua]: That's weird, that function couldn't be more perfect (in terms of runtime), but whatever.  I've fixed it.

**Comment by asamy** [comment]
Done and done.

**Pull request merged**